provider/equinix/environ.go

```diff
@@ -187,3 +199,246 @@ func (e *environ) Region() (simplestreams.CloudSpec, error) {
+			instArch = arch.ARM64
+		default:
+			continue nextPlan
+
+		}
+
+		mem, err := parseMemValue(plan.Specs.Memory.Total)
+		if err != nil {
+			continue
+		}
+
+		instTypes = append(instTypes,
+			instances.InstanceType{
+				Id:       plan.ID,
+				Name:     plan.Name,
+				CpuCores: uint64(plan.Specs.Cpus[0].Count),
+				Mem:      mem,
+				Arches:   []string{instArch},
+				// Scale per hour costs so they can be represented as an integer for sorting purposes.
+				Cost:       uint64(plan.Pricing.Hour * 1000.0),
+				Deprecated: plan.Legacy,
+			})
+	}
+
+	return instTypes, nil
+}
+
+func validPlan(plan packngo.Plan, region string) bool {
+	notAvailable := true
+	for _, a := range plan.AvailableIn {
+		if a.Code == region {
+			notAvailable = false
+			break
+		}
+	}
+	isInvalid := notAvailable || plan.Pricing == nil ||
+		plan.Specs == nil ||
+		plan.Specs.Memory == nil ||
+		len(plan.Specs.Cpus) == 0 || plan.Specs.Cpus[0].Count == 0
+
+	return !isInvalid
+}
+
+func parseMemValue(v string) (uint64, error) {
+	var scaler = uint64(1)
+	if strings.HasSuffix(v, "GB") {
+		scaler = 1024
+		v = strings.TrimSuffix(v, "GB")
+	}
+
+	val, err := strconv.ParseUint(v, 10, 64)
+	return val * scaler, err
+}
+
+func (e *environ) findInstanceSpec(controller bool, allImages []*imagemetadata.ImageMetadata, instanceTypes []instances.InstanceType, ic *instances.InstanceConstraint) (*instances.InstanceSpec, error) {
+	oss, _, err := e.equinixClient.OperatingSystems.List()
+	if err != nil {
+		return nil, err
+	}
+	var suitableImages []*imagemetadata.ImageMetadata
+
+	for _, it := range instanceTypes {
+		for _, os := range oss {
+			if !isDistroSupported(os, ic) {
+				continue
+			}
+
+			for _, p := range os.ProvisionableOn {
+				if p == it.Name {
+					image := &imagemetadata.ImageMetadata{
+						Id:   os.Slug,
+						Arch: getArchitectureFromPlan(p),
+					}
+					suitableImages = append(suitableImages, image)
+				}
+			}
+		}
+	}
+
+	images := instances.ImageMetadataToImages(suitableImages)
+	return instances.FindInstanceSpec(images, ic, instanceTypes)
+}
+
+func (e *environ) finishInstanceConfig(args *environs.StartInstanceParams, spec *instances.InstanceSpec) error {
+	matchingTools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return errors.Errorf("chosen architecture %v for image %q not present in %v",
+			spec.Image.Arch, spec.Image.Id, args.Tools.Arches())
+	}
+
+	if spec.InstanceType.Deprecated {
+		logger.Infof("deprecated instance type specified: %s", spec.InstanceType.Name)
+	}
+
+	if err := args.InstanceConfig.SetTools(matchingTools); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := instancecfg.FinishInstanceConfig(args.InstanceConfig, e.Config()); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// waitDeviceActive is a function capable of figuring out when a Equinix Metal
+// device is active
+func waitDeviceActive(c *packngo.Client, id string) (*packngo.Device, error) {
```

Comment:
An issue with this helper is that there is no way to cancel the retry loop. Given the `Attempts` and `Delay` values, you would be waiting at most around 15min if something goes wrong.  The `CloudCallContext` arg for the environ calls includes a context that you can thread through to this function to abort the loop if the user presses ctrl+c while waiting to bootstrap.

For example, you can change `IsFatalError` to:

```go
select {
  case <- ctx.Done():
    return true // treat error as fatal because we are asked to abort
  default:
    return common.IsCredentialNotValid(err)
}
```

Reply:
`CloudCallContext` does not have a `Done` function for what I can tell